kaggle_environments/envs/halite/helpers.py

```diff
@@ -0,0 +1,780 @@
+                    ship._position = wrap_point(translate_point(ship.position, ship.next_action.to_point()), configuration.size)
+                    # We don't set the new cell's ship_id here as it would be overwritten by another ship in the case of collision.
+                    # Later we'll iterate through all ships and re-set the cell._ship_id as appropriate.
+                # Clear the ship's action so it doesn't repeat the same action automatically
+                ship.next_action = None
+
+            player._halite += leftover_convert_halite
+
+        def resolve_collision(ships: List[Ship]) -> Tuple[Optional[Ship], List[Ship]]:
+            """
+            Accepts the list of ships at a particular position (must not be empty).
+            Returns the ship with the least halite or None in the case of a tie along with all other ships.
+            """
+            if len(ships) == 1:
+                return ships[0], []
+            ships_by_halite = group_by(ships, lambda ship: ship.halite)
+            smallest_halite = min(ships_by_halite.keys())
+            smallest_ships = ships_by_halite[smallest_halite]
+            if len(smallest_ships) == 1:
+                # There was a winner, return it
+                winner = smallest_ships[0]
+                return winner, [ship for ship in ships if ship != winner]
+            # There was a tie for least halite, all are deleted
+            return None, ships
+
+        # Check for ship to ship collisions
+        ship_collision_groups = group_by(board.ships.values(), lambda ship: ship.position)
+        for position, collided_ships in ship_collision_groups.items():
+            winner, deleted = resolve_collision(collided_ships)
+            if winner is not None:
+                winner.cell._ship_id = winner.id
+            for ship in deleted:
+                board._delete_ship(ship)
+                if winner is not None:
+                    # Winner takes deleted ships' halite
+                    winner._halite += ship.halite
+
+        # Check for ship to shipyard collisions
+        for shipyard in list(board.shipyards.values()):
+            ship = shipyard.cell.ship
+            if ship is None:
+                # No collision
+                continue
+            elif ship.player_id == shipyard.player_id:
+                # Ship depositing halite
+                shipyard.player._halite += ship.halite
+                ship._halite = 0
+            else:
+                # Ship to shipyard collision
+                board._delete_shipyard(shipyard)
+                board._delete_ship(ship)
+
+        # Collect halite from cells into ships
+        for ship in board.ships.values():
+            cell = ship.cell
+            delta_halite = int(cell.halite * configuration.collect_rate)
+            if cell.shipyard_id is None and delta_halite > 0:
+                ship._halite += delta_halite
+                cell._halite -= delta_halite
```

Comment:
might want another `assert cell._halite > 0` after this? Just to confirm, a cell cannot run out of halite?